x-pack/plugins/apm/server/saved_objects/migrations/migrate_legacy_apm_indices_to_space_aware.ts

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import type {
+  CoreStart,
+  Logger,
+  ISavedObjectsRepository,
+} from 'src/core/server';
+import { SavedObjectsErrorHelpers } from '../../../../../../src/core/server';
+import {
+  APM_INDEX_SETTINGS_SAVED_OBJECT_ID,
+  APM_INDEX_SETTINGS_SAVED_OBJECT_TYPE,
+} from '../../../common/apm_saved_object_constants';
+import { ApmIndicesConfig } from '../../routes/settings/apm_indices/get_apm_indices';
+import { APMIndices } from '../apm_indices';
+
+async function fetchLegacyAPMIndices(repository: ISavedObjectsRepository) {
+  try {
+    const apmIndices = await repository.get<
+      Partial<ApmIndicesConfig & { isSpaceAware: boolean }>
+    >(APM_INDEX_SETTINGS_SAVED_OBJECT_TYPE, APM_INDEX_SETTINGS_SAVED_OBJECT_ID);
+    if (apmIndices.attributes.isSpaceAware) {
+      // This has already been migrated to become space-aware
+      return null;
+    }
+    return apmIndices;
+  } catch (err) {
+    if (SavedObjectsErrorHelpers.isNotFoundError(err)) {
+      // This can happen if APM is not being used
+      return null;
+    }
+    throw err;
+  }
+}
+
+export async function migrateLegacyAPMIndicesToSpaceAware({
+  coreStart,
+  logger,
+}: {
+  coreStart: CoreStart;
+  logger: Logger;
+}) {
+  const repository = coreStart.savedObjects.createInternalRepository(['space']);
+  try {
+    // Fetch legacy APM indices
+    const legacyAPMIndices = await fetchLegacyAPMIndices(repository);
+
+    if (legacyAPMIndices === null) {
+      return;
+    }
+    // Fetch spaces available
+    const spaces = await repository.find({
+      type: 'space',
+      page: 1,
+      perPage: 10_000, // max number of spaces as of 8.2
+      fields: ['name'], // to avoid fetching *all* fields
+    });
+
+    const savedObjectAttributes: APMIndices = {
+      apmIndices: legacyAPMIndices.attributes,
+      isSpaceAware: true,
+    };
+
+    // Calls create first to update the default space setting isSpaceAware to true
+    await repository.create<
+      Partial<ApmIndicesConfig & { isSpaceAware: boolean }>
+    >(APM_INDEX_SETTINGS_SAVED_OBJECT_TYPE, savedObjectAttributes, {
+      id: APM_INDEX_SETTINGS_SAVED_OBJECT_ID,
+      overwrite: true,
+    });
+
+    // Create new APM indices space aware for all spaces available
+    await repository.bulkCreate<Partial<APMIndices>>(
+      spaces.saved_objects
+        // Skip default space since it was already updated
+        .filter(({ id: spaceId }) => spaceId !== 'default')
+        .map(({ id: spaceId }) => ({
+          id: APM_INDEX_SETTINGS_SAVED_OBJECT_ID,
+          type: APM_INDEX_SETTINGS_SAVED_OBJECT_TYPE,
+          initialNamespaces: [spaceId],
+          attributes: savedObjectAttributes,
+        }))
+    );
```

Comment:
Sorry I am commenting _after_ this PR is merged, but here goes:

You're taking the old object's whole `attributes` object and moving it to an `attributes.apmIndices` object. The problem with doing it like this is that an object that was exported from 8.1 will not be imported correctly into 8.2+.

You need to define a saved object migration for `'8.2.0'` in `x-pack/plugins/apm/server/saved_objects/apm_indices.ts` to change the attributes. That migration will always be executed _before_ Kibana starts, and it will also be executed on imported objects.

I'm happy to help you do this on a follow-up PR if you need assistancde, it's a pretty minor change.

Reply:
☹️ I'm changing it on a follow-up PR. Thanks for pointing it out.

Reply:
Great catch @jportner , thanks again. 
I thought we had test coverage for the migration. Anything we can do to catch this?